asset/asset.go

```diff
@@ -1904,38 +2117,96 @@ func (req *GroupKeyRequest) Validate() error {
 			sha256.Size)
 	}
 
+	// Version 1 specific checks.
+	if req.Version == GroupKeyV1 {
+		tapscriptRoot, err := chainhash.NewHash(req.TapscriptRoot)
+		if err != nil {
+			return fmt.Errorf("version 1 group key request " +
+				"tapscript root must be a valid hash")
+		}
+
+		if tapscriptRoot.IsEqual(&chainhash.Hash{}) {
+			return fmt.Errorf("version 1 group key request " +
+				"tapscript root must not be all zeros")
+		}
+	}
+
+	if req.ExternalKey.IsSome() && req.Version != GroupKeyV1 {
+		return fmt.Errorf("external key can only be specified for " +
+			"version 1 group key request")
+	}
+
 	return nil
 }
 
+// NewGroupPubKey derives a group key for the asset group based on the group key
+// request and the genesis asset ID.
+func (req *GroupKeyRequest) NewGroupPubKey(genesisAssetID ID) (btcec.PublicKey,
+	error) {
+
+	// If the external key is not specified, we will construct a version 0
+	// group key.
+	if req.ExternalKey.IsNone() {
+		// Compute the tweaked group key and set it in the asset before
+		// creating the virtual minting transaction.
+		groupPubKey, err := GroupPubKeyV0(
+			req.RawKey.PubKey, genesisAssetID[:], req.TapscriptRoot,
+		)
+		if err != nil {
+			return btcec.PublicKey{}, fmt.Errorf("cannot tweak "+
+				"group key: %w", err)
+		}
+
+		return *groupPubKey, nil
+	}
+
+	// At this point, the external key should be specified. We will now
+	// construct a new version 1 group key.
+	externalKey, err := req.ExternalKey.UnwrapOrErr(
+		fmt.Errorf("unexpected nil external key"),
+	)
+	if err != nil {
+		return btcec.PublicKey{}, err
+	}
+
+	groupPubKey, _, err := NewGroupKeyV1FromExternal(
+		externalKey, genesisAssetID, req.CustomTapscriptRoot,
+	)
+	if err != nil {
+		return btcec.PublicKey{}, fmt.Errorf("cannot derive group "+
+			"key: %w", err)
+	}
+
+	return groupPubKey, nil
+}
+
 // BuildGroupVirtualTx derives the tweaked group key for group key request,
 // and constructs the group virtual TX needed to construct a sign descriptor and
 // produce an asset group witness.
 func (req *GroupKeyRequest) BuildGroupVirtualTx(genBuilder GenesisTxBuilder) (
 	*GroupVirtualTx, error) {
+
 	// First, perform the final checks on the asset being authorized for
 	// group membership.
 	err := req.Validate()
 	if err != nil {
 		return nil, err
 	}
 
-	// Compute the tweaked group key and set it in the asset before
-	// creating the virtual minting transaction.
-	genesisTweak := req.AnchorGen.ID()
-	tweakedGroupKey, err := GroupPubKeyV0(
-		req.RawKey.PubKey, genesisTweak[:], req.TapscriptRoot,
-	)
+	// Construct an asset group pub key.
+	genesisAssetID := req.AnchorGen.ID()
+	groupPubKey, err := req.NewGroupPubKey(genesisAssetID)
```

Comment:
:+1: I don't mind these small commits. Your commit narrative is very clear to me.